mlflow/pyfunc/__init__.py

```diff
@@ -828,13 +828,183 @@ def _create_model_downloading_tmp_dir(should_use_nfs):
+    from mlflow.types.schema import ColSpec, TensorSpec, DataType
+    from pyspark.sql.types import ArrayType
+
+    # TODO: handle optional output columns.
+    if isinstance(spec, ColSpec):
+        return spec.type.to_spark()
+    elif isinstance(spec, TensorSpec):
+        data_type = DataType.from_numpy_type(spec.type)
+        if data_type is None:
+            raise MlflowException(
+                f"Model output tensor spec type {spec.type} is not supported in spark_udf.",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+        if len(spec.shape) == 1:
+            return ArrayType(data_type.to_spark())
+        elif len(spec.shape) == 2:
+            return ArrayType(ArrayType(data_type.to_spark()))
+        else:
+            raise MlflowException(
+                "Only one dimension or 2 dimensions tensors are supported as spark_udf "
+                f"return value, but model output '{spec.name}' has shape {spec.shape}.",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+    else:
+        raise MlflowException(
+            f"Unknown schema output spec {spec}.",
+            error_code=INVALID_PARAMETER_VALUE
+        )
+
+
+def _infer_spark_udf_return_type(model_output_schema):
+    from pyspark.sql.types import StructType, StructField, ArrayType
+
+    if len(model_output_schema.inputs) == 1:
+        return _convert_model_output_spec_to_spark_type(model_output_schema.inputs[0])
+
+    return StructType(
+        [
+            StructField(
+                name=spec.name or str(i), dataType=_convert_model_output_spec_to_spark_type(spec)
+            )
+            for i, spec in enumerate(model_output_schema.inputs)
+        ]
+    )
+
+
 def _parse_spark_datatype(datatype: str):
     from pyspark.sql.functions import udf
 
     return udf(lambda x: x, returnType=datatype).returnType
 
 
-def spark_udf(spark, model_uri, result_type="double", env_manager=_EnvManager.LOCAL):
+def _convert_array_values(values, spark_type, spark_primitive_type_to_np_type):
+    """
+    Convert list or numpy array values to spark dataframe column values.
+    """
+    from pyspark.sql.types import ArrayType
+
+    # Get element type from a array<E> type or a array<array<E>> type.
+    # and get the array_dim value (1 or 2).
+    array_dim = 1
+    elem_type = spark_type.elementType
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+        array_dim = 2
+
+    if type(elem_type) not in spark_primitive_type_to_np_type:
+        raise MlflowException(
+            "Unsupported array type field with element type "
+            f"{elem_type.simpleString()} in struct type.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+    np_type = spark_primitive_type_to_np_type[type(elem_type)]
+
+    if array_dim == 1:
+        return [np.array(v, dtype=np_type) for v in values]
+    else:
+        return [list(np.array(v, dtype=np_type)) for v in values]
```

Comment:
Hi, I'm wondering if as part of this PR, you could address this bug https://github.com/mlflow/mlflow/issues/8986 I just filed? Thank you.

Reply:
Supported this in this commit : https://github.com/mlflow/mlflow/pull/8934/commits/d8f76719dabf10f7ff5e8f248bd149849f4e97fa

Reply:
@freud14-tm We can address that in a separate PR.

Reply:
I will revert https://github.com/mlflow/mlflow/commit/d8f76719dabf10f7ff5e8f248bd149849f4e97fa and file a separate PR for it